pkg/agent/puller.go

```diff
@@ -1,73 +1,130 @@
 package agent
 
 import (
-	"fmt"
 	"github.com/kubeflow/kfserving/pkg/agent/storage"
+	v1 "github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
 	"log"
 	"path/filepath"
 )
 
+type OpType string
+
+const (
+	Add    OpType = "Add"
+	Remove OpType = "Remove"
+)
+
+// sentinel to represent model absence
+var remove v1.ModelSpec = v1.ModelSpec{}
+
 type Puller struct {
-	ChannelMap map[string]Channel
-	Downloader Downloader
+	channelMap  map[string]ModelChannel
+	completions chan string
+	Downloader  Downloader
 }
 
-type Channel struct {
-	EventChannel chan EventWrapper
+type ModelOp struct {
+	ModelName string
+	Op        OpType
+	Spec      *v1.ModelSpec
 }
 
-func (p *Puller) AddModel(modelName string) Channel {
-	// TODO: Figure out the appropriate buffer-size for this
-	// TODO: Check if event Channel exists
-	eventChannel := make(chan EventWrapper, 20)
-	channel := Channel{
-		EventChannel: eventChannel,
+func StartPuller(downloader Downloader, commands <-chan ModelOp) {
+	puller := Puller{
+		channelMap:  make(map[string]ModelChannel),
+		completions: make(chan string, 128),
+		Downloader:  downloader,
 	}
-	go p.modelProcessor(modelName, channel.EventChannel)
-	p.ChannelMap[modelName] = channel
-	return p.ChannelMap[modelName]
+	go puller.processCommands(commands)
 }
 
-func (p *Puller) RemoveModel(modelName string) error {
-	channel, ok := p.ChannelMap[modelName]
-	if ok {
-		close(channel.EventChannel)
-		delete(p.ChannelMap, modelName)
+func (p *Puller) processCommands(commands <-chan ModelOp) {
+	// channelMap accessed only by this goroutine
+	var finished bool
+	for {
+		select {
+		case modelOp, ok := <-commands:
+			if !ok {
+				finished = true
+				//commands = nil //TODO tbd
+			} else {
+				switch modelOp.Op {
+				case Add:
+					p.enqueueModelOp(modelOp.ModelName, modelOp.Spec)
+				case Remove:
+					p.enqueueModelOp(modelOp.ModelName, &remove)
+				}
+			}
+		case completed := <-p.completions:
+			p.modelOpComplete(completed, finished)
+		}
 	}
-	if err := storage.RemoveDir(filepath.Join(p.Downloader.ModelDir, modelName)); err != nil {
-		return fmt.Errorf("failing to delete model directory: %v", err)
+}
+
+type ModelChannel struct {
+	modelOps    chan *v1.ModelSpec
+	opsInFlight int
+}
+
+func (p *Puller) enqueueModelOp(modelName string, modelSpec *v1.ModelSpec) {
+	modelChan, ok := p.channelMap[modelName]
+	if !ok {
+		modelChan = ModelChannel{
+			modelOps: make(chan *v1.ModelSpec, 8),
+		}
+		go p.modelProcessor(modelName, modelChan.modelOps)
+		p.channelMap[modelName] = modelChan
+	}
+	modelChan.opsInFlight += 1
+	modelChan.modelOps <- modelSpec
+}
+
+func (p *Puller) modelOpComplete(modelName string, closed bool) {
+	modelChan, ok := p.channelMap[modelName]
+	if ok {
+		modelChan.opsInFlight -= 1
+		if modelChan.opsInFlight == 0 {
+			close(modelChan.modelOps)
+			delete(p.channelMap, modelName)
+			if closed && len(p.channelMap) == 0 {
```

Comment:
Does this mean we close the completion channel when all model events are processed? how do we create this channel again as I saw it is only created when starting the puller.

Reply:
This is just for clean teardown of the channels but I guess likely won't be needed since I assume the watcher will keep running until the container is going down anyhow.

The idea is that the ModelEvents channel would first be closed and then this logic would ensure the completions channel is closed as soon as the last in-progress/queued operation completes after that.

Reply:
I am still a little confused about why we need to close the model channel and completion channel whenn the in-flight model operation is 0. Shouldn't we only close those channels when a model is removed?

Reply:
This is my pretty much my first time writing go so I'm not sure of certain best practices etc. please don't hesitate to point out things that look odd/wrong in general :)

The idea here is to keep the goroutine and channel around only as long as it's needed to serialize multiple events that might be produced for a given model (most likely case model is removed while still in the process of being added). Once the last one of those finishes then they are cleaned up, and recreated on the next event. Otherwise if you had say 1000 models loaded, you would have the same number of channels and goroutines just sitting around doing nothing. But maybe that's not a big deal?

Reply:
I would consider not to close the channel after all in-flight events are done. I'd consider keeping it open and only close when the model is removed, as yuzhui mentioned above. Bringing up and down a channel when in-flights finish and a new one comes in doesn't seem to be most efficient. But then again, do any go experts know the cost of doing that?